Add test to better demonstrate `ActiveModel::Errors#added?` behavior

maclover7 opened the pull request:
If passing in the Symbol form identifier for the error, then it will not
be in `messages`, so we need to check `details`. Likewise, the fully
composed error message (ex: "is too long (maximum is 25 characters)") is
not in `details`, but is in `messages. Via this patch, we will now be
checking both data sources.

Fixes #25410

rafaelfranca commented:
I don't think this is a good idea. How was the behavior of this in Rails 4.2?

maclover7 commented:
[Behavior on `4-2-stable`](https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/errors.rb#L343) is very similar to that on master.

I still think that this is a good idea though -- the `errors.added?(:name, :too_long)` example from the original ticket should work IMHO.

rafaelfranca commented:
I agree it should work but not by checking `details` and `messages`. It should only check one of them.

maclover7 commented:
I totally agree. However, part of the issue here is that what this method checks accepts is slightly unclear.

Needing to support all of the following use cases is tricky 😬
- `#added?(:name, :too_long)` (checking `details`)
- `#added?(:name, :too_long, count: 25)` (checking `messages` for the error message)
- `#added?(:name, "is too long (maximum is 25 characters)")` (checking `messages` for the error message)

maclover7 commented:
Opening / closing to kick Travis build.

rafaelfranca commented:
Yea, in my opinion it should only support:

```
#added?(:name, :too_long, count: 25) (checking messages for the error message)
#added?(:name, "is too long (maximum is 25 characters)") (checking messages for the error message)
```

if the error message requires an option. `#added?(:name, :too_long, count: 24)` should return false, so `#added?(:name, :too_long)` should too since it is missing the count information.

maclover7 commented:
Hmm, I think we're uncovering a little bit of an edge case, here because when the error message does not need to interpolate any variables (ex: [this test](https://github.com/rails/rails/blob/master/activemodel/test/cases/errors_test.rb#L185-189)), then, for example, `added?(:name, :blank)` works fine. I think we should still support the `#added?(:name, :too_long)` use case mentioned above 😬

rafaelfranca commented:
I still don't think so. If we return false if the options used to build the
message is different we should return false if there is no option.
On qui, 16 de jun de 2016 at 17:19 Jon Moss <notifications@github.com>
wrote:

> Hmm, I think we're uncovering a little bit of an edge case, here because
> when the error message does not need to interpolate any variables (ex: this
> test
> <https://github.com/rails/rails/blob/master/activemodel/test/cases/errors_test.rb#L185-189>),
> then, for example, added?(:name, :blank) works fine. I think we should
> still support the #added?(:name, :too_long) use case mentioned above 😬
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/25415#issuecomment-226617031>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAC66HuQZEvOh5xKqW38TvJQrsQwsKfMks5qMb3agaJpZM4I3bEO>
> .
>

maclover7 committed:
Add test to better demonstrate `ActiveModel::Errors#added?` behavior

Fixes #25410.

maclover7 commented:
Taking another look at this, and I agree that checking both hashes is a mistake. Updated my PR to contain a regression test to better exemplify the behavior here, and removed my `lib/` changes.

maclover7 commented:
r? @rafaelfranca

rafaelfranca commented:
Thanks
